plugins/autolink/plugin.js

```diff
@@ -6,45 +6,75 @@
 ( function() {
 	'use strict';
 
-	// Regex by Imme Emosol.
-	var validUrlRegex = /^(https?|ftp):\/\/(-\.)?([^\s\/?\.#]+\.?)+(\/[^\s]*)?[^\s\.,]$/ig,
-		// Regex by (https://www.w3.org/TR/html5/forms.html#valid-e-mail-address).
-		validEmailRegex = /^[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/g,
-		doubleQuoteRegex = /"/g;
+	var doubleQuoteRegex = /"/g;
 
 	CKEDITOR.plugins.add( 'autolink', {
-		requires: 'clipboard',
+		requires: 'clipboard,textmatch',
 
 		init: function( editor ) {
+			var urlTemplate = new CKEDITOR.template( '<a href="{link}">{text}</a>' ),
+				emailTemplate = new CKEDITOR.template( '<a href="mailto:{link}">{text}</a>' );
+
 			editor.on( 'paste', function( evt ) {
-				var data = evt.data.dataValue;
+				// Pasted text may be encoded. Decode it so we will be able to compare its length with match.
+				var text = CKEDITOR.tools.htmlDecode( evt.data.dataValue );
 
-				if ( evt.data.dataTransfer.getTransferType( editor ) == CKEDITOR.DATA_TRANSFER_INTERNAL ) {
-					return;
-				}
+				// Attach afterPaste event here to avoid race condition between events.
+				editor.once( 'afterPaste', function() {
+					var matched = CKEDITOR.plugins.textMatch.match( editor.getSelection().getRanges()[ 0 ], matchCallback );
+
+					if ( !matched || matched.text.length != text.length ) {
+						return;
+					}
+
+					insertLink( matched );
+				} );
+			} );
+
+			// IE has its own link completion and we don't want to interfere with it.
+			if ( CKEDITOR.env.ie && !CKEDITOR.env.edge ) {
+				return;
+			}
+
+			editor.on( 'contentDom', function() {
+				var commitKeystrokes = editor.config.autolink_commitKeystrokes || CKEDITOR.config.autolink_commitKeystrokes;
+				editor.editable().on( 'keydown', function( evt ) {
+					if ( CKEDITOR.tools.indexOf( commitKeystrokes, evt.data.getKey() ) == -1 ) {
+						return;
+					}
 
-				// If we found "<" it means that most likely there's some tag and we don't want to touch it.
-				if ( data.indexOf( '<' ) > -1 ) {
+					var matched = CKEDITOR.plugins.textMatch.match( editor.getSelection().getRanges()[ 0 ], matchCallback );
```

Comment:
It could look something like that:

```javascript
function autolink( editor, condition ) {
	var matched = CKEDITOR.plugins.textMatch.match( editor.getSelection().getRanges()[ 0 ], matchCallback );

		if ( !condition( matched ) ) {
			return;
		}

		insertLink( matched );
}

editor.on( 'paste', function( evt ) {
	// Pasted text may be encoded. Decode it so we will be able to compare its length with match.
	var text = CKEDITOR.tools.htmlDecode( evt.data.dataValue );

	// Attach afterPaste event here to avoid race condition between events.
	editor.once( 'afterPaste', function() {
		autolink( editor, function( matched ) {
			return matched && matched.text.length === text.length;
		} )
	} );
} );
[…]
editor.on( 'contentDom', function() {
	var commitKeystrokes = editor.config.autolink_commitKeystrokes || CKEDITOR.config.autolink_commitKeystrokes;
	editor.editable().on( 'keydown', function( evt ) {
		if ( CKEDITOR.tools.indexOf( commitKeystrokes, evt.data.getKey() ) == -1 ) {
			return;
		}
		
		autolink( editor, function( matched ) {
			return matched;
		} );
		// Handle event ASAP thus some plugins may change
		// editor selection or cancel keydown events e.g. wysiwygarea, enterkey.
	}, null, null, 0 );
} );
```